File: 2015/07/07.py
variables = {}

def solver():

    global variables
    with open('./input.txt', 'r') as file:
        problem_input = [line.split() for line in file.read().split('\n')]

    to_delete = []
    for line in problem_input:
        if line[0].isdigit() and line[1] == '->':
            if line[-1] not in variables.keys():
                variables[line[-1]] = int(line[0])
            to_delete.append(line)
    for el in to_delete:
        problem_input.remove(el)

    while len(problem_input) > 0:
        to_delete = []
        for line in problem_input:
            if ((line[0].isdigit() or line[0] in variables.keys()) and 
                (line[2].isdigit() or line[2] in variables.keys())
                ):
                var_l = int(line[0]) if line[0].isdigit() else variables[line[0]]
                var_r = int(line[2]) if line[2].isdigit() else variables[line[2]]
                match line[1]:
                    case 'AND':
                        variables[line[-1]] = var_l & var_r
                        to_delete.append(line)
                    case 'OR':
                        variables[line[-1]] = var_l | var_r
                        to_delete.append(line)
                    case 'RSHIFT':
                        variables[line[-1]] = var_l >> var_r
                        to_delete.append(line)
                    case 'LSHIFT':
                        variables[line[-1]] = var_l << var_r
                        to_delete.append(line)
            if (line[0] == 'NOT' and (line[1].isdigit() or line[1] in variables.keys())):
                    var = int(line[1]) if line[1].isdigit() else variables[line[1]]
                    variables[line[-1]] = ~var
                    to_delete.append(line)
            if len(line) == 3 and line[0] in variables.keys():
                variables[line[-1]] = variables[line[0]]
                to_delete.append(line)
        for el in to_delete:
            problem_input.remove(el)

    print('a:', variables['a'] & 0xFFFF)

solver()
variables = {'b': variables['a']}
solver()

# 19.1 ms ±  6.1 ms (1000 runs, 100 warmup) - Python 3.10.5 (conda-forge)

File: 2015/07/07.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <numeric>
#include <cctype>
#include <functional>


bool string_is_digit (std::string string);
void solver (void);

std::map<std::string, unsigned short> variables {};


int main (void) {
    solver();
    unsigned short a {variables["a"]};
    variables.clear();
    variables.insert({"b", a});
    solver();
}


bool string_is_digit (std::string string) {
    return std::all_of(string.begin(), string.end(), ::isdigit);
}

void solver (void) {
    std::ifstream file("./input.txt");
    std::string line;
    std::vector<std::vector<std::string>> problem_input;

    while (std::getline(file, line)) {
        std::stringstream stream_data(line);
        std::vector<std::string> line_vector;
        std::string stream_string;
        while (std::getline(stream_data, stream_string, ' ')) {
            line_vector.push_back(stream_string);
        }
        problem_input.push_back(line_vector);
    }

    std::vector<std::vector<std::string>> to_delete;
    for (size_t i {}; i < problem_input.size(); ++i) {
        auto line {problem_input[i]};
        if (string_is_digit(line[0]) && line[1] == "->") {
            if (variables.find(line.back()) == variables.end()) {
                variables.insert({line.back(), std::stoi(line[0])});
            }
            to_delete.push_back(problem_input[i]);
        }
    }
    for (auto el : to_delete) {
        problem_input.erase(
            std::remove(problem_input.begin(), problem_input.end(), el),
            problem_input.end()
            );
    }

    while (problem_input.size() > 0) {
        to_delete.clear();

        for (size_t i {}; i < problem_input.size(); ++i) {
            auto line {problem_input[i]};

            if ((
                string_is_digit(line.at(0)) || variables.find(line.at(0)) != variables.end()
                ) && (
                string_is_digit(line.at(2)) || variables.find(line.at(2)) != variables.end()
                ))
                {
                    unsigned short var_l {
                        string_is_digit(line.at(0)) ?
                        static_cast<unsigned short>(std::stoi(line.at(0))) : 
                        variables[line.at(0)]
                        };
                    unsigned short var_r {
                        string_is_digit(line.at(2)) ? 
                        static_cast<unsigned short>(std::stoi(line.at(2))) : 
                        variables[line.at(2)]
                        };
                    switch (line[1][0])
                    {
                    case 'A':
                        variables[line.back()] = var_l & var_r;
                        to_delete.push_back(problem_input.at(i));
                        break;
                    case 'O':
                        variables[line.back()] = var_l | var_r;
                        to_delete.push_back(problem_input.at(i));
                        break;
                    case 'R':
                        variables[line.back()] = var_l >> var_r;
                        to_delete.push_back(problem_input.at(i));
                        break;
                    case 'L':
                        variables[line.back()] = var_l << var_r;
                        to_delete.push_back(problem_input.at(i));
                        break;
                    }
                }
            
            if (
                line.at(0) == "NOT"
                &&
                (string_is_digit(line.at(1)) || variables.find(line.at(1)) != variables.end())
                )
                {
                    unsigned short var {
                        string_is_digit(line[1]) ?
                        static_cast<unsigned short>(std::stoi(line.at(1))) :
                        variables[line.at(1)]
                        };
                    variables[line.back()] = ~var;
                    to_delete.push_back(problem_input.at(i));
                }
            
            if (
                line.size() == 3
                && 
                variables.find(line.at(0)) != variables.end()
                ) {
                variables[line.back()] = variables[line.at(0)];
                to_delete.push_back(problem_input.at(i));
            }
        }

        for (auto el : to_delete) {
            problem_input.erase(
                std::remove(problem_input.begin(), problem_input.end(), el),
                problem_input.end()
            );
        }
    }
    std::cout << "a: " << variables["a"] << std::endl;
}

// 4.2 ms ± 0.4 ms (1k runs, 100 warmup) - G++ 12.1.1: g++ -O3 -std=C++20